### ISSUE TITLE
Stream::merge for arbitrary number of Streams?

### ISSUE THREAD

**Issue by luser** [async-rs/async-std#693]
Hey! A while ago I was trying out `async-std` to see if I could write a nice abstraction around a thing that I had implemented in C++ in the past and found super painful: listening on multiple network addresses and handling connections from any of them. This is usually to handle IPv4+IPv6, but listening on a specific set of available IP addresses or listening on multiple ports are not uncommon things either. I love how nice the [`TcpListener::incoming`](https://docs.rs/async-std/1.4.0/async_std/net/struct.TcpListener.html#method.incoming)  API works in practice so I was trying to build something that felt the same but accepted multiple addresses.  I got something working after a while but I struggled for a bit and figured I'd open an issue here to see if there's a place for an API that would make this easier. [Here's the implementation I ended up with](https://github.com/luser/async-server-test/blob/da51eccc08139014e605e8ecdb54bacd3176a7bb/src/main.rs#L14). It uses an `mpsc::unbounded` under the hood, spawns a task for each individual `TcpListener` and has them send their streams of `Result<TcpStream, Error>` over the channel for the caller to handle.

This certainly works, but I can't help but wonder if it could be nicer if we had an API like `Stream::merge` that supported an arbitrary number of `Stream`s. Something like `stream::from_iter(...).flatten()` but yielding the first available item from any stream would make this fairly trivial to write, something like:

```rust
async fn accept_on_addrs(addrs: impl ToSocketAddrs) -> Result<impl Stream<Item = Result<TcpStream, Error>> {
  let addrs = addrs.to_socket_addrs().await?;
  Ok(stream::merge_from(stream::from_iter(addrs)
    .then(|addr| async move {
      let listener = TcpListener::bind(addr).await?;
      listener.incoming()
    })))
}
```

**Comment by yoshuawuyts** [comment]
@luser Have you seen [`Stream::merge`](https://docs.rs/async-std/1.4.0/async_std/stream/trait.Stream.html#method.merge)? Would that work for your use case?

**Comment by luser** [comment]
Yeah I did, that's why I mentioned it in the title. :wink: I think I tried using it originally but found it kind of clunky for this use case but I can't remember why. I can give it another shot and see if I can state my issues more concretely.

**Comment by luser** [comment]
Ok, I went back and gave it another try and I found out the bigger reason I wound up using that design as well as one actual rough edge. The proposal I sketched out above doesn't actually work because `TcpListener::incoming` returns a  stream that borrows from the listener, so just mapping from `TcpListener -> Incoming` won't compile. Presumably this would need async generators or something like that to be feasible.

Aside from that the part I found a little clunky is that since `Stream::merge` returns a concrete type I had to box every step along the way and also spell out the resulting type, like:
```rust
 let out: Box<dyn Stream<Item = StreamRes> + Unpin> = listeners.into_iter()
        .fold(Box::new(stream::empty()), |strm, listener| {
            Box::new(strm.merge(listener.incoming()))
        });
```

I'm not sure there's any way around this since if you have `let m1 = a.merge(b); let m2 = m1.merge(c);` `m1` and `m2` have different types, right?
